File: NextApp/Android/core/session/src/main/kotlin/jp/komehyappyo/member/next/core/session/AppSession.kt
package jp.komehyappyo.member.next.core.session

import jp.komehyappyo.member.next.core.model.UserStage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class SessionState(
    val userId: String = "guest",
    val stage: UserStage = UserStage.Guest,
    val selectedCommunityId: String? = null,
    val previousCommunityId: String? = null,
    val authenticationToken: String? = null,
)

class AppSession {
    private val mutableState = MutableStateFlow(SessionState())
    val state: StateFlow<SessionState> = mutableState.asStateFlow()

    fun updateStage(stage: UserStage, userId: String) {
        mutableState.value = mutableState.value.copy(stage = stage, userId = userId)
    }

    fun updateAuthenticatedUser(userId: String, idToken: String) {
        mutableState.value = mutableState.value.copy(
            userId = userId,
            authenticationToken = idToken,
        )
    }

    fun logout() {
        mutableState.value = SessionState()
    }

    fun selectCommunity(communityId: String?) {
        if (communityId == mutableState.value.selectedCommunityId) return
        mutableState.value = mutableState.value.copy(
            selectedCommunityId = communityId,
            previousCommunityId = mutableState.value.selectedCommunityId,
        )
    }

    fun returnToPreviousCommunity() {
        val current = mutableState.value.selectedCommunityId
        mutableState.value = mutableState.value.copy(
            selectedCommunityId = mutableState.value.previousCommunityId,
            previousCommunityId = current,
        )
    }
}

File: NextApp/iOS/OrgPortalNext/Core/Session/AppSession.swift
import Foundation
import Model

@MainActor
public final class AppSession: ObservableObject {
    @Published public private(set) var userStage: UserStage
    @Published public var selectedCommunityId: String?
    @Published public private(set) var previousCommunityId: String?
    @Published public private(set) var authenticatedUserId: String?
    public private(set) var authenticationToken: String?

    public init(
        userStage: UserStage = .guest,
        selectedCommunityId: String? = nil
    ) {
        self.userStage = userStage
        self.selectedCommunityId = selectedCommunityId
    }

    public func updateUserStage(_ stage: UserStage) {
        userStage = stage
    }

    public func updateAuthenticatedUser(userId: String, idToken: String) {
        authenticatedUserId = userId
        authenticationToken = idToken
    }

    public func logout() {
        userStage = .guest
        authenticatedUserId = nil
        authenticationToken = nil
        selectedCommunityId = nil
        previousCommunityId = nil
    }

    public func selectCommunity(_ id: String?) {
        guard id != selectedCommunityId else { return }
        previousCommunityId = selectedCommunityId
        selectedCommunityId = id
    }

    public func returnToPreviousCommunity() {
        let current = selectedCommunityId
        selectedCommunityId = previousCommunityId
        previousCommunityId = current
    }
}
